tests/functional/z_cluster/cluster_expansion/test_add_capacity.py

```diff
@@ -197,3 +208,210 @@ def test_add_capacity_pre_upgrade(self, reduce_and_resume_cluster_load):
+class TestAddCapacityStretchCluster:
+    """
+    Add capacity to the Stretch cluster with arbiter configuration
+
+    """
+
+    def test_add_capacity(
+        self,
+        setup_logwriter_cephfs_workload_factory,
+        setup_logwriter_rbd_workload_factory,
+        logreader_workload_factory,
+    ):
+        """
+        Test add capacity to the stretch cluster while the IO's
+        are running
+        """
+        sc_obj = StretchCluster()
+
+        # setup logwriter workloads in the background
+        (
+            sc_obj.cephfs_logwriter_dep,
+            sc_obj.cephfs_logreader_job,
+        ) = setup_logwriter_cephfs_workload_factory(read_duration=0)
+
+        # generate 2 minutes worth of logs
+        logger.info("Generating 2 mins worth of log")
+        time.sleep(120)
+
+        sc_obj.get_logwriter_reader_pods(label=constants.LOGWRITER_CEPHFS_LABEL)
+        sc_obj.get_logwriter_reader_pods(label=constants.LOGREADER_CEPHFS_LABEL)
+        sc_obj.get_logwriter_reader_pods(
+            label=constants.LOGWRITER_RBD_LABEL, exp_num_replicas=2
+        )
+        logger.info("All the workloads pods are successfully up and running")
+
+        start_time = datetime.now(timezone.utc)
+
+        sc_obj.get_logfile_map(label=constants.LOGWRITER_CEPHFS_LABEL)
+        sc_obj.get_logfile_map(label=constants.LOGWRITER_RBD_LABEL)
+
+        # add capacity to the cluster
+        logger.info("adding capacity to the cluster now...")
+        storage_cluster.add_capacity_lso(False)
+        logger.info("successfully added capacity to the cluster")
+
+        # check Io for any failures
+        end_time = datetime.now(timezone.utc)
+        sc_obj.post_failure_checks(start_time, end_time, wait_for_read_completion=False)
+        logger.info("Successfully verified with post failure checks for the workloads")
+
+        sc_obj.cephfs_logreader_job.delete()
+        logger.info(sc_obj.cephfs_logreader_pods)
+        for pod in sc_obj.cephfs_logreader_pods:
+            pod.wait_for_pod_delete(timeout=120)
+        logger.info("All old CephFS logreader pods are deleted")
+
+        # check for any data loss
+        assert sc_obj.check_for_data_loss(
+            constants.LOGWRITER_CEPHFS_LABEL
+        ), "[CephFS] Data is lost"
+        logger.info("[CephFS] No data loss is seen")
+        assert sc_obj.check_for_data_loss(
+            constants.LOGWRITER_RBD_LABEL
+        ), "[RBD] Data is lost"
+        logger.info("[RBD] No data loss is seen")
+
+        # check for data corruption
+        pvc = get_pvc_objs(
+            pvc_names=[
+                sc_obj.cephfs_logwriter_dep.get()["spec"]["template"]["spec"][
+                    "volumes"
+                ][0]["persistentVolumeClaim"]["claimName"]
+            ],
+            namespace=constants.STRETCH_CLUSTER_NAMESPACE,
+        )[0]
+        logreader_workload_factory(
+            pvc=pvc, logreader_path=constants.LOGWRITER_CEPHFS_READER, duration=5
+        )
```

Comment:
please check for the following, post add capacity

1. OSD count
2. Capacity
3. ceph osd tree

Reply:
completed

Reply:
for ceph osd tree verification, you should refer https://github.com/red-hat-storage/ocs-ci/blob/819a522ad93a97e41c2a0445efc7d6308c38f78f/tests/functional/z_cluster/cluster_expansion/test_add_capacity_entry_exit_criteria.py#L293

Idea here is to make sure the new disks got added under right node, rack and dc. 

Please check the functions - https://github.com/red-hat-storage/ocs-ci/blob/819a522ad93a97e41c2a0445efc7d6308c38f78f/ocs_ci/ocs/cluster.py#L1721.

Reply:
@hnallurv that is not the correct way to do it in stretch cluster as ceph osd tree output looks something like below:

```
$ ceph osd tree
ID   CLASS  WEIGHT   TYPE NAME               STATUS  REWEIGHT  PRI-AFF
 -1         1.95374  root default                                     
 -4         0.97687      zone data-1                                  
-17         0.39075          host compute-0                           
  9    ssd  0.09769              osd.9           up   1.00000  1.00000
 11    ssd  0.09769              osd.11          up   1.00000  1.00000
 14    ssd  0.09769              osd.14          up   1.00000  1.00000
 17    ssd  0.09769              osd.17          up   1.00000  1.00000
 -3         0.39075          host compute-2                           
  0    ssd  0.09769              osd.0           up   1.00000  1.00000
  5    ssd  0.09769              osd.5           up   1.00000  1.00000
 15    ssd  0.09769              osd.15          up   1.00000  1.00000
 18    ssd  0.09769              osd.18          up   1.00000  1.00000
 -7         0.19537          host compute-4                           
  2    ssd  0.09769              osd.2           up   1.00000  1.00000
  4    ssd  0.09769              osd.4           up   1.00000  1.00000
-10         0.97687      zone data-2                                  
 -9         0.39075          host compute-1                           
  1    ssd  0.09769              osd.1           up   1.00000  1.00000
  6    ssd  0.09769              osd.6           up   1.00000  1.00000
 12    ssd  0.09769              osd.12          up   1.00000  1.00000
 19    ssd  0.09769              osd.19          up   1.00000  1.00000
-13         0.39075          host compute-3                           
  3    ssd  0.09769              osd.3           up   1.00000  1.00000
  7    ssd  0.09769              osd.7           up   1.00000  1.00000
 13    ssd  0.09769              osd.13          up   1.00000  1.00000
 16    ssd  0.09769              osd.16          up   1.00000  1.00000
-15         0.19537          host compute-5                           
  8    ssd  0.09769              osd.8           up   1.00000  1.00000
 10    ssd  0.09769              osd.10          up   1.00000  1.00000
``` 
Each of the osd have a weight assigned to it. Each zone has weight calculated based out of sum of the weights of osds in each zone. If there is some imbalance in the osds between zone then the weight value for each zone will be different. if its balanced then the weight value will be equal. This is the logic I'm using in my code to make sure osds are equally distributed.

Reply:
The goal of the function you mentioned doesnt hold good for stretch cluster configuration. You can see the docstring.

```python
def check_osd_tree_1az_vmware(osd_tree, number_of_osds):
    """
    Checks whether an OSD tree is created/modified correctly. This can be used as a verification step for
    deployment and cluster expansion tests.
    This function is specifically for ocs cluster created on 1 AZ VMWare setup

    Args:
        osd_tree (dict): Dictionary of the values which represent 'osd tree'.
        number_of_osds (int): total number of osds in the cluster

    Returns:
        bool: True, if the ceph osd tree is formed correctly. Else False

    """
    # in case of vmware, there will be only one zone as of now. The OSDs are arranged as follows:
    # ID  CLASS WEIGHT  TYPE NAME                            STATUS REWEIGHT PRI-AFF
    # -1       0.99326 root default
    # -8       0.33109     rack rack0
    # -7       0.33109         host ocs-deviceset-0-0-dktqc
    #  1   hdd 0.33109             osd.1                        up  1.00000 1.00000
    # There will be 3 racks - rack0, rack1, rack2.
    # When cluster expansion is successfully done, a host and an osd are added in each rack.
    # The number of hosts will be equal to the number osds the cluster has. Each rack can
    # have multiple hosts but each host will have only one osd under it.
```

Reply:
Confirmed about this with Travis:
```
Travis Nielsen
  [6:52 PM](https://ibm-systems-storage.slack.com/archives/D06RJ2YPK8F/p1713360149309429)
Hi Mahesh
[6:52](https://ibm-systems-storage.slack.com/archives/D06RJ2YPK8F/p1713360167114639)
Correct, if the weights match across the zones, then the OSDs are equally weighted
[6:53](https://ibm-systems-storage.slack.com/archives/D06RJ2YPK8F/p1713360180190799)
The weights of individual OSDs are based on the size of the device
[6:53](https://ibm-systems-storage.slack.com/archives/D06RJ2YPK8F/p1713360210139289)
So as long as all the OSDs are the same size, then their sum will match for the zones. If any OSD is a different size, the weights might not exactly match
```
@hnallurv